src/bindgen/language_backend/clike.rs

```diff
@@ -1009,4 +1036,611 @@ impl LanguageBackend for CLikeLanguageBackend<'_> {
+        if let Some(condition) = &some_condition {
+            condition.write_before(self.config, out);
+        }
+
+        match d {
+            ItemContainer::Struct(ref s) => write!(out, "struct {};", s.export_name),
+            ItemContainer::Union(ref u) => write!(out, "union {};", u.export_name),
+            ItemContainer::Enum(ref e) => {
+                if self.config.language == Language::C {
+                    write!(out, "enum {};", e.export_name);
+                } else {
+                    unreachable!();
+                }
+            }
+            _ => unreachable!(),
+        }
+
+        if let Some(condition) = &some_condition {
+            condition.write_after(self.config, out);
+        }
+    }
+}
+
+/// State of an item.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum ResolveOrderState {
+    Pending,            // cannot be used
+    Declaration(usize), // declared, can be used in pointers and references
+    Alias(Path),        // declared, defined when path is defined
+    Defined,            // defined (and declared), can be used
+}
+
+/// What the resolver wants.
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub enum ResolveOrderWants {
+    Declaration(Path),
+    Definition(Path),
+}
+impl ResolveOrderWants {
+    pub fn path(&self) -> &Path {
+        match self {
+            Self::Declaration(ref path) => path,
+            Self::Definition(ref path) => path,
+        }
+    }
+}
+
+/// Resolve the order of dependencies.
+///
+/// Can try to preserve the original order.
+/// Assumes all unknown paths are defined.
+/// Multiple items can have the same path.
+///
+/// Algorithm:
+/// - if no pending items: done
+/// - if countdown is over or no item: give up
+/// - if no unresolved dependencies: alias or define item, reset countdown or retry pending
+/// - if can declare dependencies: declare dependencies, alias or define item, reset countdown or retry pending
+/// - next item, decrease countdown
+pub struct ResolveOrder<'a> {
+    /// Configuration.
+    pub config: &'a Config,
+    /// Contains items that need to be resolved.
+    pub dependencies: &'a mut Dependencies,
+    /// Location of each path.
+    pub paths2indexes: MultiMap<Path, usize>,
+    /// Contains the state of each item.
+    pub states: Vec<ResolveOrderState>,
+    /// The pending order of items.
+    pub pending: Vec<usize>,
+    /// Current pending item being processed.
+    pub current: usize,
+    /// When to give up.
+    pub countdown: Option<usize>,
+    /// The resolved order of items.
+    pub order: Vec<usize>,
+}
+
+impl<'a> ResolveOrder<'a> {
+    /// Create a new resolver capable of changing the order of the dependencies.
+    ///
+    /// Does not try to preserve the original order.
+    pub fn new(config: &'a Config, dependencies: &'a mut Dependencies) -> Self {
+        let num_items = dependencies.order.len();
+        let mut paths2indexes: MultiMap<Path, usize> = MultiMap::new();
```

Comment:
Rather than pulling a new dependency for this, I think we can just use a `HashMap<Path, Vec<usize>>` right?